sdk/dotnet/Pulumi.Tests/Deployment/DeploymentRunnerTests.cs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.Logging;
+using Xunit;
+
+using Pulumi.Testing;
+using Pulumi.Tests.Mocks;
+
+namespace Pulumi.Tests
+{
+    public class DeploymentRunnerTests
+    {
+        [Fact]
+        public async Task WorksUnderStress()
+        {
+            var resources = await Deployment.TestAsync<StressRunnerStack>(new EmptyMocks());
+            var stack = (StressRunnerStack)resources[0];
+            var result = await stack.RunnerResult;
+            Assert.Equal(0, result);
+        }
+
+        class StressRunnerStack : Stack
+        {
+            public Task<int> RunnerResult { get; private set; }
+            public StressRunnerStack()
+            {
+                var runner = Pulumi.Deployment.Instance.Internal.Runner;
+
+                for (var i = 0; i < 100; i++)
+                {
+                    runner.RegisterTask($"task{i}", Task.Delay(100 + i));
+                }
+
+                this.RunnerResult = runner.RunAsync<EmptyStack>();
```

Comment:
This test failed on master. This is our first test to reproduce the original problem. It was non-deterministic but N=100 was enough to trigger it pretty consistently. Consider it a regression test?  I an add a comment about that.

The `RunAsync<EmptyStack>()` clause indirectly invokes `WhileRunningAsync()` which is where the bug was. Should I instead make that internal and invoke that directly?

Reply:
Ok this is pretty interesting I investigated a bit further here. 

The `RunAsync<EmptyStack>` is essential to the repro. In fact it seems that I can only repro the exception when more than one stack is being deployed against the same Runner object (and same Deployment object). So to reproduce the issue we need a race of two concurrent `WhileRunningAsync()` threads on the same runner. In fact even with these changes here, we will not see exceptions, but we might have race conditions in this case of mis-attributing exceptions (or spurious waits). If the single runner is servicing two stacks, it's going to mixup their exceptions.

So taking a big step back. Do we believe it's an error to be running two stacks simultaneously against the single Deployment? If it's an error, how can we best prevent that from happening?

Reply:
I missed this on my review. To @mikhailshilkov 's point I didn't even know this was possible. I don't think a consumer would be able to do this because the access modifiers would prevent them from accessing the Runner.

Reply:
Conceptually I think it is an error. Currently a Deployment is an instance of a pulumi update action against a single stack, and the Runner is an implementation detail of that Deployment - and it is 1:1. I don't think it would ever make sense the way this it is currently architected to have 2 update actions executing against a single Deployment.

Reply:
So the issue is that we need to find a way to repro the original issue which was a flake race condition caused by access to the inner task dictionary of the Runner? It's always difficult to try to repro a flakey issue but make we could instantiate a Runner for testing purposes and just throw N tasks at it? Maybe you don't need a Deployment to repro since this issue was isolated to the Runner

Reply:
Hrm. That could just be a bad test case. Maybe this situation is unrealistic but under some other situations this condition can arise as well. I'm leaning a little to accepting changes here as they make the locking a little more foolproof?

Let me have one more look at CI failures that manifested with the exception we're fixing here.

Also. We could put a little check into WhileRunningAsync to make sure it's never invoked concurrently (one instance at a time), and fail loudly if it does. If we believe it should never happen, but if we manage to do it somehow anyway, that'd catch it.

Reply:
Probably wouldn't hurt anything. Yea when I made that original Automation API PR I specifically refactored the core SDK so that all pulumi execution enters [right here](https://github.com/pulumi/pulumi/blob/81f87b0c6ac5a2b52268b46599affdf79d9e6db7/sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs#L186), even the Automation API enters there. And you'll see the first thing it does is instantiate a Deployment

Reply:
I've put a PR here with the defensive check https://github.com/pulumi/pulumi/pull/7597 - locally it does not seem to detect it, let's see on CI... Also I locally sporadically get "key not found" in the tests, without triggering the concurrent WRA. So basically I have a poor repro test here, there must be other ways to repro the situation.

Reply:
Yeah that PR basically passes! So far so good.